File: repository/serieDAO.js
import conectar from "./Conexao.js";
import Serie from "../model/serie.js";

export default class SerieDAO {

    constructor() {
        this.init();
    }

    async init() {
        try {
            const conexao = await conectar.connect();
            const sql = `
                CREATE TABLE IF NOT EXISTS serie (
                    serie_id SERIAL PRIMARY KEY,
                    serie_num INT NOT NULL,
                    serie_descr VARCHAR NOT NULL
                );
            `;
            await conexao.query(sql);
            conexao.release();
        } catch (erro) {
            console.log("Erro ao iniciar tabela serie: " + erro.message);
        }
    }

    async gravar(serie) {
        if (serie instanceof Serie) {
            const conexao = await conectar.connect();
            const sql = `INSERT INTO serie (serie_num, serie_descr) VALUES ($1, $2)`;
            const parametros = [serie.getSerieNum(), serie.getDescricao()];
            await conexao.query(sql, parametros);
            conexao.release();
        }
    }

    async alterar(serie) {
        if (serie instanceof Serie) {
            const conexao = await conectar.connect();
            const sql = `UPDATE serie SET serie_num = $1, serie_descr = $2 WHERE serie_id = $3`;
            const parametros = [serie.getSerieNum(), serie.getDescricao(), serie.getId()];
            await conexao.query(sql, parametros);
            conexao.release();
        }
    }

    async apagar(id) {
        const conexao = await conectar.connect();
        const sql = `DELETE FROM serie WHERE serie_id = $1`;
        const parametros = [id];
        await conexao.query(sql, parametros);
        conexao.release();
    }

    async get(filtro) {
        const conexao = await conectar.connect();
        let sql;
        let parametros;
        let unico = false;

        if (!filtro) {
            sql = `SELECT * FROM serie`;
            parametros = [];
        } 
        else if (!isNaN(filtro)) {
            sql = `SELECT * FROM serie WHERE serie_id = $1`;
            parametros = [filtro];
            unico = true;
        } 
        else {
            sql = `SELECT * FROM serie WHERE serie_num::TEXT LIKE $1 OR serie_descr ILIKE $1`;
            parametros = [`%${filtro}%`];
        }

        const result = await conexao.query(sql, parametros);
        conexao.release();

        if (result.rows.length > 0) {
            const listaSeries = result.rows.map(row =>
                new Serie(row.serie_num, row.serie_descr, row.serie_id)
            );
            return unico ? listaSeries[0] : listaSeries;
        }
        return null;
    }
}

File: src/main/java/sistema/saga/sagabackend/repository/SerieDAO.java
package sistema.saga.sagabackend.repository;

import org.springframework.stereotype.Repository;
import sistema.saga.sagabackend.model.Serie;

import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

@Repository
public class SerieDAO {

    public boolean gravar(Serie serie, Conexao conexao) {
        String sql = """
        INSERT INTO serie (serie_num, serie_descr)
        VALUES (#1, '#2')
    """;
        sql = sql.replace("#1", String.valueOf(serie.getSerieNum()));
        sql = sql.replace("#2", serie.getSerieDescr().replace("'", "''"));

        System.out.println("SQL executado: " + sql); // <--- debug aqui

        return conexao.manipular(sql);
    }

    public boolean alterar(Serie serie, Conexao conexao) {
        String sql = """
            UPDATE serie 
            SET serie_num = '#1', serie_descr = '#2'
            WHERE serie_id = '#3'
        """;
        sql = sql.replace("#1", String.valueOf(serie.getSerieNum()));
        sql = sql.replace("#2", serie.getSerieDescr());
        sql = sql.replace("#3", String.valueOf(serie.getSerieId()));
        return conexao.manipular(sql);
    }

    public boolean apagar(Serie serie, Conexao conexao) {
        String sql = "DELETE FROM serie WHERE serie_id = '#1'";
        sql = sql.replace("#1", String.valueOf(serie.getSerieId()));
        return conexao.manipular(sql);
    }

    public int getSerie(Serie serie, Conexao conexao) {
        String sql = "SELECT * FROM serie WHERE serie_id = '#1'";
        sql = sql.replace("#1", String.valueOf(serie.getSerieId()));

        try {
            ResultSet rs = conexao.consultar(sql);
            if (rs.next()) {
                serie.setSerieId(rs.getInt("serie_id"));
                serie.setSerieNum(rs.getInt("serie_num"));
                serie.setSerieDescr(rs.getString("serie_descr"));
                return rs.getInt("serie_id");
            }
        } catch (Exception e) {
            throw new RuntimeException("Erro ao buscar série por ID", e);
        }

        return 0;
    }

    public List<Serie> get(String filtro, Conexao conexao) {
        List<Serie> series = new ArrayList<>();
        String sql;

        if (filtro == null || filtro.isEmpty()) {
            sql = "SELECT * FROM serie ORDER BY serie_num";
        }
        else {
            sql = "SELECT * FROM serie WHERE serie_descr ILIKE '%" + filtro + "%' OR CAST(serie_num AS TEXT) LIKE '%" + filtro + "%'";
        }

        try {
            ResultSet rs = conexao.consultar(sql);
            while (rs.next()) {
                Serie serie = new Serie(
                        rs.getInt("serie_id"),
                        rs.getInt("serie_num"),
                        rs.getString("serie_descr")
                );
                series.add(serie);
            }
        } catch (Exception e) {
            throw new RuntimeException("Erro ao consultar séries", e);
        }

        return series;
    }
}
